Can't implement InternalSaveChangesDsl for another backend

## Setup

### Versions

- **Rust:** nightly-2018-03-07-x86_64-unknown-linux-gnu
- **Diesel:** 1.2.2
- **Database:** Oracle 12g
- **Operating System** Ubuntu 18.04

## Problem Description
I am trying to get another backend for diesel working. Namely oracle/oci. Therefore I need to implement e.g. ```InternalSaveChangesDsl```. Since all oracle/oci related code is currently outside of diesel, it is impossible to implement this. Please note that I made `save_changes_dsl` public.

### What is the actual output?
```
error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g. `MyStruct<T>`); only traits defined in the current crate can be implemented for a type parameter
  --> src/oracle/query_dsl/save_changes_dsl.rs:15:1
   |
15 | / impl<T, U> InternalSaveChangesDsl<OciConnection, U> for T
16 | |     where
17 | |         T: Copy + Identifiable,
18 | |         T: AsChangeset<Target = <T as HasTable>::Table> + IntoUpdateTarget,
...  |
26 | |     }
27 | | }
   | |_^
```
### Steps to reproduce
There is some code at https://github.com/GiGainfosystems/diesel-oci/tree/feature/save_changes_dsl available. I also had to change some diesel code: https://github.com/GiGainfosystems/diesel/tree/feature/save_changes_dsl
Within `diesel/diesel_tests` I run
```
cargo +nightly test --features oracle
```

## Checklist

- [X] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) for similar issues.
- [X] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)

--- a/diesel/src/query_dsl/mod.rs
+++ b/diesel/src/query_dsl/mod.rs
@@ -15,8 +15,8 @@
 
 use backend::Backend;
 use connection::Connection;
-use expression::Expression;
 use expression::count::CountStar;
+use expression::Expression;
 use helper_types::*;
 use query_builder::locking_clause as lock;
 use query_source::{joins, Table};

--- a/diesel/src/query_dsl/mod.rs
+++ b/diesel/src/query_dsl/mod.rs
@@ -15,8 +15,8 @@
 
 use backend::Backend;
 use connection::Connection;
-use expression::count::CountStar;
 use expression::Expression;
+use expression::count::CountStar;
 use helper_types::*;
 use query_builder::locking_clause as lock;
 use query_source::{joins, Table};
@@ -48,9 +48,7 @@ pub use self::group_by_dsl::GroupByDsl;
 pub use self::join_dsl::{InternalJoinDsl, JoinOnDsl, JoinWithImplicitOnClause};
 #[doc(hidden)]
 pub use self::load_dsl::LoadQuery;
-pub use self::save_changes_dsl::SaveChangesDsl;
-#[doc(hidden)]
-pub use self::save_changes_dsl::InternalSaveChangesDsl;
+pub use self::save_changes_dsl::{SaveChangesDsl, UpdateAndFetchResults};
 
 /// The traits used by `QueryDsl`.
 ///

--- a/diesel/src/query_dsl/mod.rs
+++ b/diesel/src/query_dsl/mod.rs
@@ -49,6 +49,8 @@ pub use self::join_dsl::{InternalJoinDsl, JoinOnDsl, JoinWithImplicitOnClause};
 #[doc(hidden)]
 pub use self::load_dsl::LoadQuery;
 pub use self::save_changes_dsl::SaveChangesDsl;
+#[doc(hidden)]
+pub use self::save_changes_dsl::InternalSaveChangesDsl;
 
 /// The traits used by `QueryDsl`.
 ///

--- a/diesel/src/query_dsl/save_changes_dsl.rs
+++ b/diesel/src/query_dsl/save_changes_dsl.rs
@@ -115,10 +115,11 @@ where
 /// #     Ok(())
 /// # }
 /// ```
-pub trait SaveChangesDsl<Conn>: Sized {
+pub trait SaveChangesDsl<Conn> {
     /// See the trait documentation.
     fn save_changes<T>(self, connection: &Conn) -> QueryResult<T>
     where
+        Self: Sized,
         Conn: InternalSaveChangesDsl<Self, T>,
     {
         connection.internal_save_changes(self)

--- a/diesel/src/query_dsl/save_changes_dsl.rs
+++ b/diesel/src/query_dsl/save_changes_dsl.rs
@@ -12,21 +12,22 @@ use query_dsl::methods::{ExecuteDsl, FindDsl};
 use query_dsl::{LoadQuery, RunQueryDsl};
 use result::QueryResult;
 
-pub trait InternalSaveChangesDsl<Conn, T>: Sized {
-    fn internal_save_changes(self, connection: &Conn) -> QueryResult<T>;
+#[doc(hidden)]
+pub trait InternalSaveChangesDsl<T, U> {
+    fn internal_save_changes(&self, changeset: T) -> QueryResult<U>;
 }
 
 #[cfg(feature = "postgres")]
 use pg::PgConnection;
 
 #[cfg(feature = "postgres")]
-impl<T, U> InternalSaveChangesDsl<PgConnection, U> for T
+impl<T, U> InternalSaveChangesDsl<T, U> for PgConnection
 where
     T: Copy + AsChangeset<Target = <T as HasTable>::Table> + IntoUpdateTarget,
     Update<T, T>: LoadQuery<PgConnection, U>,
 {
-    fn internal_save_changes(self, conn: &PgConnection) -> QueryResult<U> {
-        ::update(self).set(self).get_result(conn)
+    fn internal_save_changes(&self, changeset: T) -> QueryResult<U> {
+        ::update(changeset).set(changeset).get_result(self)
     }
 }
 
@@ -34,7 +35,7 @@ where
 use sqlite::SqliteConnection;
 
 #[cfg(feature = "sqlite")]
-impl<T, U> InternalSaveChangesDsl<SqliteConnection, U> for T
+impl<T, U> InternalSaveChangesDsl<T, U> for SqliteConnection
 where
     T: Copy + Identifiable,
     T: AsChangeset<Target = <T as HasTable>::Table> + IntoUpdateTarget,
@@ -42,9 +43,9 @@ where
     Update<T, T>: ExecuteDsl<SqliteConnection>,
     Find<T::Table, T::Id>: LoadQuery<SqliteConnection, U>,
 {
-    fn internal_save_changes(self, conn: &SqliteConnection) -> QueryResult<U> {
-        try!(::update(self).set(self).execute(conn));
-        T::table().find(self.id()).get_result(conn)
+    fn internal_save_changes(&self, changeset: T) -> QueryResult<U> {
+        try!(::update(changeset).set(changeset).execute(self));
+        T::table().find(changeset.id()).get_result(self)
     }
 }
 
@@ -52,7 +53,7 @@ where
 use mysql::MysqlConnection;
 
 #[cfg(feature = "mysql")]
-impl<T, U> InternalSaveChangesDsl<MysqlConnection, U> for T
+impl<T, U> InternalSaveChangesDsl<T, U> for MysqlConnection
 where
     T: Copy + Identifiable,
     T: AsChangeset<Target = <T as HasTable>::Table> + IntoUpdateTarget,
@@ -60,9 +61,9 @@ where
     Update<T, T>: ExecuteDsl<MysqlConnection>,
     Find<T::Table, T::Id>: LoadQuery<MysqlConnection, U>,
 {
-    fn internal_save_changes(self, conn: &MysqlConnection) -> QueryResult<U> {
-        try!(::update(self).set(self).execute(conn));
-        T::table().find(self.id()).get_result(conn)
+    fn internal_save_changes(&self, changeset: T) -> QueryResult<U> {
+        try!(::update(changeset).set(changeset).execute(self));
+        T::table().find(changeset.id()).get_result(self)
     }
 }
 
@@ -114,13 +115,13 @@ where
 /// #     Ok(())
 /// # }
 /// ```
-pub trait SaveChangesDsl<Conn> {
+pub trait SaveChangesDsl<Conn>: Sized {
     /// See the trait documentation.
     fn save_changes<T>(self, connection: &Conn) -> QueryResult<T>
     where
-        Self: InternalSaveChangesDsl<Conn, T>,
+        Conn: InternalSaveChangesDsl<Self, T>,
     {
-        self.internal_save_changes(connection)
+        connection.internal_save_changes(self)
     }
 }
 

--- a/diesel/src/query_dsl/save_changes_dsl.rs
+++ b/diesel/src/query_dsl/save_changes_dsl.rs
@@ -19,21 +19,25 @@ use result::QueryResult;
 /// The only case where it is required to work with this trait is while
 /// implementing a new connection type.
 /// Otherwise use [`SaveChangesDsl`](trait.SaveChangesDsl.html)
-pub trait UpdateAndFetchResults<T, U>: Connection {
+///
+/// For implementing this trait for a custom backend:
+/// * The `Changes` generic parameter represents the changeset that should be stored
+/// * The `Output` generic parameter represents the type of the response.
+pub trait UpdateAndFetchResults<Changes, Output>: Connection {
     /// See the traits documentation.
-    fn update_and_fetch(&self, changeset: T) -> QueryResult<U>;
+    fn update_and_fetch(&self, changeset: Changes) -> QueryResult<Output>;
 }
 
 #[cfg(feature = "postgres")]
 use pg::PgConnection;
 
 #[cfg(feature = "postgres")]
-impl<T, U> UpdateAndFetchResults<T, U> for PgConnection
+impl<Changes, Output> UpdateAndFetchResults<Changes, Output> for PgConnection
 where
-    T: Copy + AsChangeset<Target = <T as HasTable>::Table> + IntoUpdateTarget,
-    Update<T, T>: LoadQuery<PgConnection, U>,
+    Changes: Copy + AsChangeset<Target = <Changes as HasTable>::Table> + IntoUpdateTarget,
+    Update<Changes, Changes>: LoadQuery<PgConnection, Output>,
 {
-    fn update_and_fetch(&self, changeset: T) -> QueryResult<U> {
+    fn update_and_fetch(&self, changeset: Changes) -> QueryResult<Output> {
         ::update(changeset).set(changeset).get_result(self)
     }
 }
@@ -42,17 +46,17 @@ where
 use sqlite::SqliteConnection;
 
 #[cfg(feature = "sqlite")]
-impl<T, U> UpdateAndFetchResults<T, U> for SqliteConnection
+impl<Changes, Output> UpdateAndFetchResults<Changes, Output> for SqliteConnection
 where
-    T: Copy + Identifiable,
-    T: AsChangeset<Target = <T as HasTable>::Table> + IntoUpdateTarget,
-    T::Table: FindDsl<T::Id>,
-    Update<T, T>: ExecuteDsl<SqliteConnection>,
-    Find<T::Table, T::Id>: LoadQuery<SqliteConnection, U>,
+    Changes: Copy + Identifiable,
+    Changes: AsChangeset<Target = <Changes as HasTable>::Table> + IntoUpdateTarget,
+    Changes::Table: FindDsl<Changes::Id>,
+    Update<Changes, Changes>: ExecuteDsl<SqliteConnection>,
+    Find<Changes::Table, Changes::Id>: LoadQuery<SqliteConnection, Output>,
 {
-    fn update_and_fetch(&self, changeset: T) -> QueryResult<U> {
+    fn update_and_fetch(&self, changeset: Changes) -> QueryResult<Output> {
         try!(::update(changeset).set(changeset).execute(self));
-        T::table().find(changeset.id()).get_result(self)
+        Changes::table().find(changeset.id()).get_result(self)
     }
 }
 
@@ -60,17 +64,17 @@ where
 use mysql::MysqlConnection;
 
 #[cfg(feature = "mysql")]
-impl<T, U> UpdateAndFetchResults<T, U> for MysqlConnection
+impl<Changes, Output> UpdateAndFetchResults<Changes, Output> for MysqlConnection
 where
-    T: Copy + Identifiable,
-    T: AsChangeset<Target = <T as HasTable>::Table> + IntoUpdateTarget,
-    T::Table: FindDsl<T::Id>,
-    Update<T, T>: ExecuteDsl<MysqlConnection>,
-    Find<T::Table, T::Id>: LoadQuery<MysqlConnection, U>,
+    Changes: Copy + Identifiable,
+    Changes: AsChangeset<Target = <Changes as HasTable>::Table> + IntoUpdateTarget,
+    Changes::Table: FindDsl<Changes::Id>,
+    Update<Changes, Changes>: ExecuteDsl<MysqlConnection>,
+    Find<Changes::Table, Changes::Id>: LoadQuery<MysqlConnection, Output>,
 {
-    fn update_and_fetch(&self, changeset: T) -> QueryResult<U> {
+    fn update_and_fetch(&self, changeset: Changes) -> QueryResult<Output> {
         try!(::update(changeset).set(changeset).execute(self));
-        T::table().find(changeset.id()).get_result(self)
+        Changes::table().find(changeset.id()).get_result(self)
     }
 }
 

--- a/diesel/src/query_dsl/save_changes_dsl.rs
+++ b/diesel/src/query_dsl/save_changes_dsl.rs
@@ -1,6 +1,7 @@
 use associations::HasTable;
 #[cfg(any(feature = "sqlite", feature = "mysql"))]
 use associations::Identifiable;
+use connection::Connection;
 #[cfg(any(feature = "sqlite", feature = "mysql"))]
 use dsl::Find;
 #[cfg(any(feature = "sqlite", feature = "postgres", feature = "mysql"))]
@@ -12,21 +13,27 @@ use query_dsl::methods::{ExecuteDsl, FindDsl};
 use query_dsl::{LoadQuery, RunQueryDsl};
 use result::QueryResult;
 
-#[doc(hidden)]
-pub trait InternalSaveChangesDsl<T, U> {
-    fn internal_save_changes(&self, changeset: T) -> QueryResult<U>;
+/// A trait defining how to update a record and fetch the updated entry
+/// on a certain backend.
+///
+/// The only case where it is required to work with this trait is while
+/// implementing a new connection type.
+/// Otherwise use [`SaveChangesDsl`](trait.SaveChangesDsl.html)
+pub trait UpdateAndFetchResults<T, U>: Connection {
+    /// See the traits documentation.
+    fn update_and_fetch(&self, changeset: T) -> QueryResult<U>;
 }
 
 #[cfg(feature = "postgres")]
 use pg::PgConnection;
 
 #[cfg(feature = "postgres")]
-impl<T, U> InternalSaveChangesDsl<T, U> for PgConnection
+impl<T, U> UpdateAndFetchResults<T, U> for PgConnection
 where
     T: Copy + AsChangeset<Target = <T as HasTable>::Table> + IntoUpdateTarget,
     Update<T, T>: LoadQuery<PgConnection, U>,
 {
-    fn internal_save_changes(&self, changeset: T) -> QueryResult<U> {
+    fn update_and_fetch(&self, changeset: T) -> QueryResult<U> {
         ::update(changeset).set(changeset).get_result(self)
     }
 }
@@ -35,7 +42,7 @@ where
 use sqlite::SqliteConnection;
 
 #[cfg(feature = "sqlite")]
-impl<T, U> InternalSaveChangesDsl<T, U> for SqliteConnection
+impl<T, U> UpdateAndFetchResults<T, U> for SqliteConnection
 where
     T: Copy + Identifiable,
     T: AsChangeset<Target = <T as HasTable>::Table> + IntoUpdateTarget,
@@ -43,7 +50,7 @@ where
     Update<T, T>: ExecuteDsl<SqliteConnection>,
     Find<T::Table, T::Id>: LoadQuery<SqliteConnection, U>,
 {
-    fn internal_save_changes(&self, changeset: T) -> QueryResult<U> {
+    fn update_and_fetch(&self, changeset: T) -> QueryResult<U> {
         try!(::update(changeset).set(changeset).execute(self));
         T::table().find(changeset.id()).get_result(self)
     }
@@ -53,7 +60,7 @@ where
 use mysql::MysqlConnection;
 
 #[cfg(feature = "mysql")]
-impl<T, U> InternalSaveChangesDsl<T, U> for MysqlConnection
+impl<T, U> UpdateAndFetchResults<T, U> for MysqlConnection
 where
     T: Copy + Identifiable,
     T: AsChangeset<Target = <T as HasTable>::Table> + IntoUpdateTarget,
@@ -61,7 +68,7 @@ where
     Update<T, T>: ExecuteDsl<MysqlConnection>,
     Find<T::Table, T::Id>: LoadQuery<MysqlConnection, U>,
 {
-    fn internal_save_changes(&self, changeset: T) -> QueryResult<U> {
+    fn update_and_fetch(&self, changeset: T) -> QueryResult<U> {
         try!(::update(changeset).set(changeset).execute(self));
         T::table().find(changeset.id()).get_result(self)
     }
@@ -120,9 +127,9 @@ pub trait SaveChangesDsl<Conn> {
     fn save_changes<T>(self, connection: &Conn) -> QueryResult<T>
     where
         Self: Sized,
-        Conn: InternalSaveChangesDsl<Self, T>,
+        Conn: UpdateAndFetchResults<Self, T>,
     {
-        connection.internal_save_changes(self)
+        connection.update_and_fetch(self)
     }
 }
 

